pkg/apis/meta.go

```diff
@@ -0,0 +1,303 @@
+	SetCreatedBy(user string)
+	GetUpdatedBy() string
+	SetUpdatedBy(user string)
+	GetFolder() string
+	SetFolder(uid string)
+	GetSlug() string
+	SetSlug(v string)
+	GetTitle() string
+	SetTitle(v string)
+	GetOriginInfo() (*ResourceOriginInfo, error)
+	SetOriginInfo(info *ResourceOriginInfo)
+	GetOriginName() string
+	GetOriginPath() string
+	GetOriginKey() string
+	GetOriginTimestamp() (*time.Time, error)
+}
+
+var _ GrafanaResourceMetaAccessor = (*grafanaResourceMetaAccessor)(nil)
+
+type grafanaResourceMetaAccessor struct {
+	obj runtime.Object
+}
+
+func MetaAccessor(obj runtime.Object) GrafanaResourceMetaAccessor {
+	return &grafanaResourceMetaAccessor{obj}
+}
+
+func (m *grafanaResourceMetaAccessor) set(key string, val string) {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return
+	}
+
+	anno := acc.GetAnnotations()
+	if val == "" {
+		if anno != nil {
+			delete(anno, key)
+		}
+	} else {
+		if anno == nil {
+			anno = make(map[string]string)
+		}
+		anno[key] = val
+	}
+	acc.SetAnnotations(anno)
+}
+
+func (m *grafanaResourceMetaAccessor) get(key string) string {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return ""
+	}
+
+	return acc.GetAnnotations()[key]
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedTimestamp() (*time.Time, error) {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return nil, err
+	}
+
+	v, ok := acc.GetAnnotations()[AnnoKeyUpdatedTimestamp]
+	if !ok || v == "" {
+		return nil, nil
+	}
+	t, err := time.Parse(time.RFC3339, v)
+	if err != nil {
+		return nil, fmt.Errorf("invalid updated timestamp: %s", err.Error())
+	}
+	return &t, nil
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestampMillis(v int64) {
+	if v > 0 {
+		t := time.UnixMilli(v)
+		m.SetUpdatedTimestamp(&t)
+	} else {
+		m.SetUpdatedTimestamp(nil)
+	}
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestamp(v *time.Time) {
+	txt := ""
+	if v != nil {
+		txt = v.UTC().Format(time.RFC3339)
+	}
+	m.set(AnnoKeyUpdatedTimestamp, txt)
+}
+
+func (m *grafanaResourceMetaAccessor) GetCreatedBy() string {
+	return m.get(AnnoKeyCreatedBy)
+}
+
+func (m *grafanaResourceMetaAccessor) SetCreatedBy(user string) {
+	m.set(AnnoKeyCreatedBy, user)
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedBy() string {
+	return m.get(AnnoKeyUpdatedBy)
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedBy(user string) {
+	m.set(AnnoKeyUpdatedBy, user)
+}
+
+func (m *grafanaResourceMetaAccessor) GetFolder() string {
+	return m.get(AnnoKeyFolder)
+}
+
+func (m *grafanaResourceMetaAccessor) SetFolder(uid string) {
+	m.set(AnnoKeyFolder, uid)
+}
+
+func (m *grafanaResourceMetaAccessor) GetSlug() string {
+	return m.get(AnnoKeySlug)
+}
+
+func (m *grafanaResourceMetaAccessor) SetSlug(v string) {
+	m.set(AnnoKeySlug, v)
+}
+func (m *grafanaResourceMetaAccessor) GetTitle() string {
+	// look for Spec.Title or Spec.Name
+	r := reflect.ValueOf(m.obj)
+	if r.Kind() == reflect.Ptr || r.Kind() == reflect.Interface {
```

Comment:
This is interesting and very confusing - can you add some comments that explain the cases in which the object would be a pointer vs interface vs struct? I'd be worried about refactoring any of this without understanding why we have so many possible inputs.

Reply:
basically because we're accepting a k8s runtime.Object interface, so what we get could well be a pointer to a struct, or a bare struct. I forget why we test for interface as well, that might not be necessary.